compiler/rustc_mir/src/transform/check_unsafety.rs

```diff
@@ -232,37 +232,49 @@ impl<'a, 'tcx> Visitor<'tcx> for UnsafetyChecker<'a, 'tcx> {
                     UnsafetyViolationKind::GeneralAndConstFn,
                     UnsafetyViolationDetails::DerefOfRawPointer,
                 ),
-                ty::Adt(adt, _) => {
-                    if adt.is_union() {
-                        if context == PlaceContext::MutatingUse(MutatingUseContext::Store)
-                            || context == PlaceContext::MutatingUse(MutatingUseContext::Drop)
-                            || context == PlaceContext::MutatingUse(MutatingUseContext::AsmOutput)
-                        {
-                            let elem_ty = match elem {
-                                ProjectionElem::Field(_, ty) => ty,
-                                _ => span_bug!(
-                                    self.source_info.span,
-                                    "non-field projection {:?} from union?",
-                                    place
-                                ),
-                            };
-                            if !elem_ty.is_copy_modulo_regions(
+                ty::Adt(adt, _) if adt.is_union() => {
+                    let assign_to_field = matches!(
+                        context,
+                        PlaceContext::MutatingUse(
+                            MutatingUseContext::Store
+                                | MutatingUseContext::Drop
+                                | MutatingUseContext::AsmOutput
+                        )
+                    );
+                    // If there is a `Deref` further along the projection chain, this is *not* an
+                    // assignment to a union field. In that case the union field is just read to
+                    // obtain the pointer/reference.
```

Comment:
Yeah, right-to-left iteration probably makes more sense. I tried to change the existing code as little as possible, and also I couldn't find good helper methods for this reverse traversal, which was surprising enough that I wondered if it really is the right choice.

Reply:
There kinda is IIRC, but it's not great. IMO what might make sense is first (before the loop over the projection elements) finding the *last* `Deref`, if any, and iterating forwards on the rest to handle borrows / assignments.

Here that approach is complicated by the fact that `union` field *reads* still need to be checked in the "pointer" part of the projection, and we would want to avoid duplication.

IMO ideally we'd have a `PlaceBase` that's (at least semantically) either a `Local` or a dereference of one, but that's a refactor that I think dropped off the radar and won't happen any time soon.

Reply:
@RalfJung minimal diffs are always admirable =) I would though be happy to see this refactored to do more of a rightward traversal. At some point I was advocating for a pattern of using [`PlaceRef`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/struct.PlaceRef.html) so that we could write recursive processing of places more naturally. The idea would be to make a `PlaceRef` from the place and then to split the projections into `(..x, tail)` essentially and process `tail`, then create a new placeref with `x` as the slice of projections (and process it recursively). I still think this pattern would be fairly elegant and would fit perfectly here.

What @eddyb suggested could also work, I guess I just prefer to think of things as a "recursive tree" in cases like this, rather than thinking of them in "list" form.

That said, I think that to do what I wanted elegantly, it would involve adding a few helpers to `PlaceRef` (e.g., a way to split into the right-most projection and the remaining `PlaceRef`).

Reply:
I can do the change to a reverse `PlaceRef` change, but I think I'd go for an iterator instead of a "tail" split.

However, this will be a major refactor of this code, and also touch e.g. the packed field check (which seems wrong currently, it might even have false negatives). Seems a bit strange to mix up refactoring, bugfix, and intended behavioral change all in one PR.